signature-specification.md

```diff
@@ -0,0 +1,215 @@
+1. Compute the Base64Url value of ProtectedHeaders.
+2. Compute the Base64Url value of JWSPayload.
+3. Build message to be signed by concatenating the values generated in step 1 and step 2 using '.'
+`ASCII(BASE64URL(UTF8(ProtectedHeaders)) ‘.’ BASE64URL(JWSPayload))`
+4. Compute the signature on the message constructed in the previous step by using the signature algorithm defined in the corresponding header element: alg.
+5. Compute the Base64Url value of the signature produced in the previous step. This is the value of the signature property used in the signature envelope.
+
+**Signature Envelope**:
+The final signature envelope comprises of Claims, ProtectedHeaders, UnprotectedHeaders, and signature.
+
+Since Notary v2 restricts one signature per signature envelope, the compliant signature envelope MUST be in flattened JWS JSON format.
+```json
+{
+   "payload": "<Base64Url(JWSPayload)>",
+   "protected": "<Base64Url(ProtectedHeaders)>",
+   "header": {
+     "timestamp": "<Base64Url(TimeStampToken)>",
+     "x5c": ["<Base64(DER(leafCert))>", "<Base64(DER(intermediateCACert))>", "<Base64(DER(rootCert))>"]
+   },
+   "signature": "Base64Url( sign( ASCII( <Base64Url(ProtectedHeader)>.<Base64Url(JWSPayload)> )))"  
+}
+```
+
+**Implementation Constraints**
+Notary v2 implementation MUST enforce the following constraints on signature generation and verification:
+1. `alg` header value MUST NOT be `none` or any symmetric-key algorithm such as `HMAC`.
+2. `alg` header value MUST be same as that of signature algorithm identified using signing certificate's public key algorithm and size.
+3. `alg` header values for various signature algorithms:
+   | Signature Algorithm    | "alg" Param Value |
+   | --------    | -------- |
+   | ECDSA on secp256r1 with SHA-256 | ES256     |
+   | ECDSA on secp256r1 with SHA-512 | ES256     |
+   | RSASSA-PSS with SHA-256   | PS256     |
+   | RSASSA-PSS with SHA-512   | PS512     |
+4. Signing certificate MUST be a valid codesigning certificate.
+5. Only JWS JSON flattened format is supported. See 'Signature Envelope' section.
+
+## Signature Algorithms
+The implementation MUST support the following set of algorithms:
+1. RSASSA-PSS with SHA-256
+1. RSASSA-PSS with SHA-384
+1. RSASSA-PSS with SHA-512
+1. ECDSA on secp256r1 with SHA-256
+1. ECDSA on secp384r1 with SHA-384
+1. ECDSA on secp256r1 with SHA-512
+
+
+### Algorithm Selection
+The signing certificate's public key algorithm and size MUST be used to determine the signature algorithm.
+
+|Public Key Algorithm| Key Size (bits)   | Signature Algorithm  |
+| -------- | --------    | -------- |
+|RSA| 2048 | RSASSA-PSS with SHA-256     |
+|RSA| 3072 | RSASSA-PSS with SHA-384     |
+|RSA| 4096 | RSASSA-PSS with SHA-512     |
+|ECDSA| 256   | ECDSA on secp256r1 with SHA-256     |
+|ECDSA| 384   | ECDSA on secp384r1 with SHA-384     |
+|ECDSA| 512   | ECDSA on secp521r1 with SHA-512     |
```

Comment:
Pedantically, `ECDSA` is not a public key algorithm where `DSA` stands for **Digital Signature Algorithm**. Therefore, we should use `EC` here for **Elliptic Curve** keys.